akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ShardingProducerController.scala

```diff
@@ -0,0 +1,429 @@
+        ref,
+        askReplyTo => DurableProducerQueue.LoadState[A](askReplyTo)) {
+        case Success(s) => LoadStateReply(s)
+        case Failure(_) => LoadStateFailed(attempt) // timeout
+      }
+    }
+  }
+
+}
+
+class ShardingProducerController[A: ClassTag](
+    context: ActorContext[ShardingProducerController.InternalCommand],
+    producerId: String,
+    producer: ActorRef[ShardingProducerController.RequestNext[A]],
+    msgAdapter: ActorRef[ShardingEnvelope[A]],
+    region: ActorRef[ShardingEnvelope[ConsumerController.SequencedMessage[A]]],
+    durableQueue: Option[ActorRef[DurableProducerQueue.Command[A]]]) {
+  import ShardingProducerController._
+  import DurableProducerQueue.StoreMessageSent
+  import DurableProducerQueue.StoreMessageSentAck
+  import DurableProducerQueue.StoreMessageConfirmed
+  import DurableProducerQueue.MessageSent
+
+  private val requestNextAdapter: ActorRef[ProducerController.RequestNext[A]] =
+    context.messageAdapter(WrappedRequestNext.apply)
+
+  private def active(s: State[A]): Behavior[InternalCommand] = {
+
+    def onMessage(
+        entityId: EntityId,
+        msg: A,
+        replyTo: Option[ActorRef[Done]],
+        totalSeqNr: TotalSeqNr,
+        newReplyAfterStore: Map[TotalSeqNr, ActorRef[Done]]): Behavior[InternalCommand] = {
+
+      val outKey = s"$producerId-$entityId"
+      val newState =
+        s.out.get(outKey) match {
+          case Some(out @ OutState(_, _, Some(nextTo), _, _, _)) =>
+            // there is demand, send immediately
+            send(msg, outKey, out.seqNr, nextTo)
+            val newUnconfirmed = out.unconfirmed :+ Unconfirmed(totalSeqNr, out.seqNr, replyTo)
+            s.copy(
+              out = s.out.updated(outKey, out.copy(seqNr = out.seqNr + 1, nextTo = None, unconfirmed = newUnconfirmed)),
+              replyAfterStore = newReplyAfterStore)
+          case Some(out @ OutState(_, _, None, buffered, _, _)) =>
+            // no demand, buffer
+            // FIXME limit the buffers.
+            context.log.info("Buffering message to entityId [{}], buffer size [{}]", entityId, buffered.size + 1)
+            val newBuffered = buffered :+ Buffered(totalSeqNr, msg, replyTo)
+            val newS =
+              s.copy(
+                out = s.out.updated(outKey, out.copy(buffered = newBuffered)),
+                replyAfterStore = newReplyAfterStore)
+            // send an updated RequestNext to indicate buffer usage
+            producer ! createRequestNext(newS)
+            newS
+          case None =>
+            context.log.info("Creating ProducerController for entity [{}]", entityId)
+            val send: ConsumerController.SequencedMessage[A] => Unit = { seqMsg =>
+              region ! ShardingEnvelope(entityId, seqMsg)
+            }
+            val p = context.spawn(ProducerController[A](outKey, durableQueueBehavior = None, send), entityId)
+            p ! ProducerController.Start(requestNextAdapter)
```

Comment:
I can't see we ever GC these, in a sharded scenario isn't there a risk we end up with one PC per sharded entity after a while, since requests for any entity could come to "our" node?

Reply:
yes, I'll add FIXME about cleanup of unused

Reply:
Maybe we can watch the corresponding ConsumerController, to be notified of entity passivation. It should still survive a rebalance, so some idle time aspect is still needed.